privacy-on-beam/pbeam/aggregations.go

```diff
@@ -450,6 +481,176 @@ func clampNegativePartitionsFloat64Fn(v beam.V, r float64) (beam.V, float64) {
+	for k := range a.PartitionMap {
+		b.PartitionMap[k] = true
+	}
+	return b
+}
+
+func (fn *PartitionsMapFn) ExtractOutput(m mapAccum) PMap {
+	return m.PartitionMap
+}
+
+// dropValue turns a PCollection<K,V> to a PCollection<V>.
+func dropValue(key beam.X, value beam.V) beam.X {
+	return key
+}
+
+// prunePartitionsVFn takes a PCollection<K, V> as input, and returns a
+// PCollection<K, V>, where unspecified partitions have been dropped.
+// Used for count and distinct_id.
+type prunePartitionsVFn struct {
+	PartitionType beam.EncodedType
+	partitionEnc  beam.ElementEncoder
+}
+
+func newPrunePartitionsVFn(partitionType typex.FullType) *prunePartitionsVFn {
+	return &prunePartitionsVFn{
+		PartitionType: beam.EncodedType{partitionType.Type()},
+	}
+}
+
+func (fn *prunePartitionsVFn) Setup() {
+	fn.partitionEnc = beam.NewElementEncoder(fn.PartitionType.T)
+}
+
+func (fn *prunePartitionsVFn) ProcessElement(id beam.X, partitionKey beam.V, partitionsIter func(*PMap) bool, emit func(beam.X, beam.V)) {
+	var partitionBuf bytes.Buffer
+	if err := fn.partitionEnc.Encode(partitionKey, &partitionBuf); err != nil {
+		log.Exitf("pbeam.prunePartitionsVFn.ProcessElement: couldn't encode partition %v: %v", partitionKey, err)
+	}
+	var partitionMap PMap
+	partitionsIter(&partitionMap)
+	if partitionMap != nil {
+		encodedPartitionKey := string(partitionBuf.Bytes())
+		if partitionMap[encodedPartitionKey] {
+			emit(id, partitionKey)
+		}
+	}
+}
+
+// prunePartitionsFn takes a PCollection<ID, kv.Pair{K,V}> as input, and returns a
+// PCollection<ID, kv.Pair{K,V}>, where unspecified partitions have been dropped.
+// Used for sum and mean.
+func prunePartitionsKVFn(id beam.X, pair kv.Pair, partitionsIter func(*PMap) bool, emit func(beam.X, kv.Pair)) {
+	var partitionMap PMap
+	partitionsIter(&partitionMap)
+	if partitionMap != nil {
+		// Parameters in a kv.Pair are already encoded.
+		encodedPartitionKey := string(pair.K)
+		if partitionMap[encodedPartitionKey] {
+			emit(id, pair)
+		}
+	}
+}
+
+// outputPartitionsNotInTheDataFn emits partitions that are specified but not in the data.
+type outputPartitionsNotInTheDataFn struct {
+	PartitionType beam.EncodedType
+	partitionEnc  beam.ElementEncoder
+}
+
+func newOutputPartitionsNotInTheDataFn(partitionType typex.FullType) *outputPartitionsNotInTheDataFn {
+	return &outputPartitionsNotInTheDataFn{
+		PartitionType: beam.EncodedType{partitionType.Type()},
+	}
+}
+
+func (fn *outputPartitionsNotInTheDataFn) Setup() {
+	fn.partitionEnc = beam.NewElementEncoder(fn.PartitionType.T)
+}
+
+func (fn *outputPartitionsNotInTheDataFn) ProcessElement(partitionKey beam.X, value beam.V, partitionsIter func(*PMap) bool, emit func(beam.X, beam.V)) {
+	var partitionBuf bytes.Buffer
+	if err := fn.partitionEnc.Encode(partitionKey, &partitionBuf); err != nil {
+		log.Exitf("pbeam.preparePruneFn.ProcessElement: couldn't encode partition %v: %v", partitionKey, err)
+	}
+	var partitionMap PMap
+	partitionsIter(&partitionMap)
```

Comment:
Sure, I mentioned above that this function doesn't have a return value... So I log.Exitf'd instead.

Reply:
ahhh in this function partitionMap could very well be empty (nil) and it would be working right... because it could be the case that there are no specified partitions already in the data..

Reply:
Sorry, I forgot to mention this.